feat: Validate Data options and Data values

gavindsouza opened the pull request:
**DocType Validations:**

![Screenshot 2020-03-25 at 3 12 04 PM](https://user-images.githubusercontent.com/36654812/77527515-42aba280-6eb2-11ea-8f1b-6eaf3179c3bb.png)

**Document Validations:**

![Screenshot 2020-03-25 at 4 20 12 PM](https://user-images.githubusercontent.com/36654812/77528799-87383d80-6eb4-11ea-890c-e48e40871528.png)

**Client Side Validation:**

![2020-03-25 16 22 49](https://user-images.githubusercontent.com/36654812/77529265-4b51a800-6eb5-11ea-8592-ee6340e6df9f.gif)

**Server Side Validation:**

![2020-03-25 16 24 01](https://user-images.githubusercontent.com/36654812/77529282-51478900-6eb5-11ea-9aa6-c1a036572de7.gif)


Other fixes:
 - fix: ValidationError message for "{email_id} is not a valid Email Address"

[Docs Link](https://github.com/frappe/erpnext_com/pull/766)

gavindsouza committed:
feat: add validate_phone api to window to validate phone numbers

gavindsouza committed:
feat: server validations for data field options

gavindsouza committed:
feat: added server validations for phone and email

for fieldtype "Data" and options set to "Email" or "Phone", updating
    documents will trigger data validations for email and phone numbers will be made

gavindsouza committed:
feat: validations for Email and Phone Data options

added set_invalid control for docfields. On event of docfield.invalid is
set, in a given form, set_invalid will be executed which changes the
textbox colour to red

gavindsouza committed:
chore: remove deprecated validate_email_add function

gavindsouza committed:
style: rename variable emph => email_invalid

gavindsouza committed:
fix: remove type annotations

gavindsouza committed:
fix: safe check of options in docfield

gavindsouza committed:
chore: update .eslintrc globals

style: updated data.js spacing format

surajshetty3416 commented on frappe/core/doctype/doctype/doctype.py:
```diff
@@ -942,6 +942,17 @@ def scrub_fetch_from(field):
+				docfield_label = frappe.bold(docfield.label)
+				data_field_str = "<br><ul><li>" + "</li><li>".join(data_field_options) + "</ul>"
+				text = "{0} is an Invalid Data field.{1} Only Options allowed for Data field are: {2}"
+				message = _(text).format(docfield_label, "<br><br>", data_field_str)
```
Invalid translation syntax

https://frappe.io/docs/user/en/translations#1-literal-strings

surajshetty3416 commented on frappe/core/doctype/doctype/doctype.py:
```diff
@@ -942,6 +942,17 @@ def scrub_fetch_from(field):
+				message = _(text).format(docfield_label, "<br><br>", data_field_str)
+				frappe.msgprint(message, raise_exception=True)
+
+
```
Remove extra newlines

gavindsouza replied:
2 line spaces were added for consistency?

surajshetty3416 replied:
We are following the convention of leaving 2 line spaces after a class for visual separation of a class method and other functions. 1 line between functions...

surajshetty3416 commented on frappe/public/js/frappe/form/controls/data.js:
```diff
@@ -87,56 +87,29 @@ frappe.ui.form.ControlData = frappe.ui.form.ControlInput.extend({
 		return val==null ? "" : val;
 	},
 	validate: function(v) {
+		if (!v){
```
```suggestion
		if (!v) {
```

surajshetty3416 commented on frappe/utils/__init__.py:
```diff
@@ -75,11 +75,18 @@ def extract_email_id(email):
+		return False
+
+	phone_number = phone_number.strip()
+	match = re.match("([0-9\ \+\_\-\,\.\*\#\(\)]){1,20}$", phone_number)
```
```suggestion
	match = re.match("([0-9\ \+\_\-\,\.\*\#\(\)]){3,20}$", phone_number)
```

Alteast 3 characters for a valid phone number?

gavindsouza replied:
There are some countries with 2 digit emergency numbers too. We'd rather just keep this flexible?

refs: https://en.wikipedia.org/wiki/List_of_emergency_telephone_numbers

surajshetty3416 replied:
2 then?

surajshetty3416 commented on frappe/public/js/frappe/utils/datatype.js:
```diff
@@ -44,6 +44,10 @@ window.validate_email = function(txt) {
 	return frappe.utils.validate_type(txt, "email");
 };
 
+window.validate_phone = function(txt) {
```
Add `validate_phone` to .eslintrc's globals

gavindsouza committed:
fix: Improper ValidationError string

returning empty string if parseaddr returns None shows " is not a valid Email Address"

eg: if 'email@' is passed as value, email.utils.parseaddr will return a
(None, None) tuple

gavindsouza committed:
fix: dont validate email if owner and email_data is in STANDARD_USERS

gavindsouza committed:
fix: translatable string for data field options

gavindsouza committed:
style: remove extra spaces for consistency

surajshetty3416 committed:
Merge branch 'develop' into data-options

gavindsouza committed:
chore: update msgprint title

fix: circular import

gavindsouza committed:
Merge branch 'data-options' into data-options

surajshetty3416 committed:
Merge branch 'develop' into data-options

gavindsouza committed:
test(DocType): added test_data_field_options

gavindsouza committed:
tests: updated tests to cover more data field options

surajshetty3416 commented on frappe/core/doctype/doctype/test_doctype.py:
```diff
@@ -114,20 +114,29 @@ def test_all_depends_on_fields_conditions(self):
+					# assert that only data options in frappe.model.data_field_options are valid
+					self.assertRaises(frappe.ValidationError, test_doctype.insert)
+				else:
+					test_valid_doctype.insert()
```
```suggestion
					test_doctype.insert()
```

gavindsouza committed:
tests: simplifying test_data_field_options

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

surajshetty3416 committed:
style: Fix indentation

gavindsouza commented:
In submitted documents, if "Allow on submit" is enabled for certain fields, when the document is validated again, in case the Data fields throw ValidationError, the user may not be able to update those fields. This feature will stay in `develop` for this reason and will not be backported to previous versions.
